src/NuGet.Core/NuGet.CommandLine.XPlat/Commands/PackageReferenceCommands/AddPackageReferenceCommandRunner.cs

```diff
@@ -0,0 +1,332 @@
+
+                var restoreContext = new RestoreArgs()
+                {
+                    CacheContext = cacheContext,
+                    LockFileVersion = LockFileFormat.Version,
+                    Log = packageReferenceArgs.Logger,
+                    MachineWideSettings = new XPlatMachineWideSetting(),
+                    GlobalPackagesFolder = packageReferenceArgs.PackageDirectory,
+                    PreLoadedRequestProviders = providers,
+                    Sources = packageReferenceArgs.Sources?.ToList()
+                };
+
+                // Generate Restore Requests. There will always be 1 request here since we are restoring for 1 project.
+                var restoreRequests = await RestoreRunner.GetRequests(restoreContext);
+
+                // Run restore without commit. This will always return 1 Result pair since we are restoring for 1 request.
+                var restoreResult = await RestoreRunner.RunWithoutCommit(restoreRequests, restoreContext);
+
+                return restoreResult.Single();
+            }
+        }
+
+        private static async Task<DependencyGraphSpec> GetProjectDependencyGraphAsync(
+            PackageReferenceArgs packageReferenceArgs,
+            string dgFilePath,
+            int timeOut)
+        {
+            var dotnetLocation = packageReferenceArgs.DotnetPath;
+
+            if (!File.Exists(dotnetLocation))
+            {
+                throw new Exception(
+                    string.Format(CultureInfo.CurrentCulture, Strings.Error_DotnetNotFound));
+            }
+            var argumentBuilder = new StringBuilder($@" /t:GenerateRestoreGraphFile");
+
+            // Set the msbuild verbosity level if specified
+            var msbuildVerbosity = Environment.GetEnvironmentVariable(NUGET_RESTORE_MSBUILD_VERBOSITY);
+
+            if (string.IsNullOrEmpty(msbuildVerbosity))
+            {
+                argumentBuilder.Append(" /v:q ");
+            }
+            else
+            {
+                argumentBuilder.Append($" /v:{msbuildVerbosity} ");
+            }
+
+            // Pass dg file output path
+            argumentBuilder.Append(" /p:RestoreGraphOutputPath=");
+            AppendQuoted(argumentBuilder, dgFilePath);
+
+            packageReferenceArgs.Logger.LogInformation($"{dotnetLocation} msbuild {argumentBuilder.ToString()}");
+
+            var processStartInfo = new ProcessStartInfo
+            {
+                UseShellExecute = false,
+                FileName = dotnetLocation,
+                WorkingDirectory = Path.GetDirectoryName(packageReferenceArgs.ProjectPath),
+                Arguments = $"msbuild {argumentBuilder.ToString()}",
+                RedirectStandardError = true,
+                RedirectStandardOutput = true
+            };
+
+            packageReferenceArgs.Logger.LogDebug($"{processStartInfo.FileName} {processStartInfo.Arguments}");
+
+            using (var process = Process.Start(processStartInfo))
```

Comment:
Can dotnet.exe pass a delegate that allows this method to call dotnet msbuild directly instead? Shelling out seems wrong here.